types/wlr_idle.c

```diff
@@ -36,13 +39,18 @@ static void handle_activity(struct wlr_idle_timeout *timer) {
 	if (!timer->enabled) {
 		return;
 	}
-	// rearm the timer
-	wl_event_source_timer_update(timer->idle_source, timer->timeout);
+
 	// in case the previous state was sleeping send a resume event and switch state
 	if (timer->idle_state) {
 		timer->idle_state = false;
 		org_kde_kwin_idle_timeout_send_resumed(timer->resource);
 	}
+
+	// rearm the timer
+	wl_event_source_timer_update(timer->idle_source, timer->timeout);
+	if (timer->timeout == 0) {
+		idle_notify(timer);
```

Comment:
How will the wayland marshalling handle if the client destroys the timer in the earlier `send_resumed` message?
(I guess you do that in the swayidle PR and it doesn't blow up, but am a bit curious - I guess the client only sends a request to destroy and we'll destroy later so it'll work out on the server side, but the client side would receive an idle event on something it thinks doesn't exist anymore? Or does the wayland marshalling recognize it got destroyed and just discard it?)

Reply:
In `swayidle`, I destroy the client-side object in the resumed handler. The server sends resume immediately followed by idle. libwayland handles this well and discards the idle event.